mobile/src/main/java/ch/epfl/sdp/mobile/ui/play/ExpandableFloatingActionButton.kt

```diff
@@ -0,0 +1,240 @@
+ * change depending on whether the button is currently expanded or not.
+ */
+interface ExpandableFloatingActionButtonColors {
+
+  /**
+   * Returns a state with the color for the button background.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun background(expanded: Boolean): State<Color>
+
+  /**
+   * Returns a state with the color for the button border.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun border(expanded: Boolean): State<Color>
+
+  /**
+   * Returns a state with the color for the button content color.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun contentColor(expanded: Boolean): State<Color>
+}
+
+/**
+ * A composable which displays a floating action which may be expanded.
+ *
+ * @param expandedContent the content to display when expanded. Usually a list of
+ * [ExpandableFloatingActionButtonItem].
+ * @param modifier the [Modifier] for this composable.
+ * @param state the backing state for this button.
+ * @param colors the colors for this button.
+ * @param elevation the elevation for this button.
+ * @param interactionSource the [MutableInteractionSource] to keep track of inputs.
+ * @param expandedCornerSize the size of the rounded corners when the button is expanded.
+ * @param collapsedContent the content to display when collapsed.
+ */
+@OptIn(
+    ExperimentalAnimationApi::class,
+    ExperimentalMaterialApi::class,
+)
+@Composable
+fun ExpandableFloatingActionButton(
+    expandedContent: @Composable ColumnScope.() -> Unit,
+    modifier: Modifier = Modifier,
+    state: ExpandableFloatingActionButtonState = rememberExpandableFloatingActionButtonState(),
+    colors: ExpandableFloatingActionButtonColors = DefaultExpandableFloatingActionButtonColors,
+    elevation: ButtonElevation = ButtonDefaults.elevation(),
+    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
+    expandedCornerSize: Dp = 8.dp,
+    collapsedContent: @Composable RowScope.() -> Unit,
+) {
+  val color by colors.background(state.expanded)
+  val border by colors.border(state.expanded)
+  val contentColor by colors.contentColor(state.expanded)
+  val progress = animateFloatAsState(if (state.expanded) 1f else 0f)
+  val corner = remember(expandedCornerSize) { ProgressiveCornerSize(expandedCornerSize, progress) }
+
+  Surface(
+      color = color,
+      contentColor = contentColor,
+      border = BorderStroke(2.dp, border),
+      elevation = elevation.elevation(enabled = true, interactionSource).value,
+      shape = RoundedCornerShape(corner),
+      modifier = modifier,
+  ) {
+    AnimatedContent(
+        targetState = state.expanded,
+        contentAlignment = Alignment.Center,
+        transitionSpec = { FadeEnterTransition with FadeExitTransition },
+    ) { expanded ->
+      ProvideTextStyle(MaterialTheme.typography.button) {
```

Comment:
[Question] Just to make sure that my understanding is correct.  Does the `ProvideTextStyle` composable make sure that everything under its scope will be using the predefined button typography in `MaterialTheme` ? If so why don't we use the default defined style in `PawniesTheme` ?

Reply:
`ProvideTextStyle` assigns a default `TextStyle` to the `CompositionLocal` used by default by `Text()` composables. When we apply a `PawniesTheme`, we essentially set some values to all the Material Design typography styles, including `MaterialTheme.typography.button`. Therefore, if we retrieve this style later in the composition, it's populated with our `PawniesTheme` typography.

However, this also means we could use the `ExpandableFloatingActionButton` if we're not in a `PawniesTheme` but in another composition which has some other Material Design typography, and the composable would adapt and texts look good by default.